Fix autosnippets

carbon-steel opened the pull request:
Hey @L3MON4D3, this PR is in regards to https://github.com/L3MON4D3/LuaSnip/issues/706.

You'll have to forgive my code and inexperience. I've only ever contributed to C++ codebases. I believe the tests were meant for an Ubuntu environment. I am running arch so I'll have to figure out how to run them tomorrow (unless you know what I should do?).

I'd be happy to hear any notes you may have :)

L3MON4D3 commented:
Oh, I'm also running them on arch, they should work as-is.
Could you outline why this works?
Does it work if only a single character is inserted?

L3MON4D3 commented:
Oh and a big thank you already for taking a look, awesome! :)

carbon-steel commented:
> Oh, I'm also running them on arch, they should work as-is.

Ah, I got confused because of the use of apt-get in the github workflows.

> Could you outline why this works?

[b:changedtick](https://neovim.io/doc/user/eval.html#b%3Achangedtick) is a monotonically increasing per-buffer count of the number of changes made to the buffer.

You can observe its behavior by using:
`:autocmd TextChangedI * echom b:changedtick`

Then type some stuff on the buffer you called the command from. You can then see how `b:changedtick` changes with each character you insert by looking at the printed results:
`:messages`

You can see how `b:changedtick` gets incremented by one with each character you type in. If you yank some string with multiple characters and you paste it (using <C-r> in insert mode), you'll notice that `b:changedtick` increases by the number of characters that were pasted. However, the TextChangedI event only occurs once after the paste is complete. By this, we can distinguish when a TextChangedI event was triggered by single character from a normal keypress vs multiple characters from <C-r>. I use this to prevent autosnippets from expanding when text was pasted using <C-r>.

> Does it work if only a single character is inserted?

No, this will not work if you paste a single character with <C-r> (I checked) since b:changedtick's behavior will be the same as if you had typed that single character in yourself.

carbon-steel commented:
I believe remapping `<C-r>` to just use the paste command from the vim.api could fix this problem for all cases:
```
local function insertModePaste()
  local register = vim.cmd([[
  let s:register = getchar()
  echo  s:register
  ]], true)
  vim.api.nvim_paste(vim.fn.getreg(register), true, -1)
end

map('i', '<C-r>', insertModePaste)
```

I don't know if you think this should go in the plugin, but users could certainly use this code to completely fix this issue for themselves for all cases.

L3MON4D3 commented:
Thanks for the clarification👍
Seems like an okay solution, it's a bit involved, and doesn't work for all cases (if the first action is such a paste, luasnip_changedtick would be nil and we would auto-expand (not 100% sure this is possible but it might be)), but it is better that the status quo.

Changing C-R seems dangerous 😅😅

AFAICT ultisnips could have had this same issue, would you be interested in figuring out how they handled it?

carbon-steel commented:
I've fixed the test that I broke and I added a test to document and preserve the new behavior. I've been using Luasnip with my changes and it seems to work well. I think this pull request is ready to go.

What do you think about adding the code from [my above post](https://github.com/L3MON4D3/LuaSnip/pull/709#issuecomment-1368021442) to the wiki to let people know they can fully fix this issue for themselves if the add it to their configs?

carbon-steel commented:
I just tested the pasting triggers in insert mode with ultisnips and their autosnippets just expand in every case. It looks like they did not solve this problem.

L3MON4D3 commented:
Thank you for taking a look at ultisnips, in that case this seems to be the best way to do it

> I've fixed the test that I broke and I added a test to document and preserve the new behavior. I've been using Luasnip with my changes and it seems to work well. I think this pull request is ready to go.
> 
> What do you think about adding the code from [my above post](https://github.com/L3MON4D3/LuaSnip/pull/709#issuecomment-1368021442) to the wiki to let people know they can fully fix this issue for themselves if the add it to their configs?

Yes! Would you mention that in the commit-message as well? Or put it into the doc (since it fixes an issue, and is not just a new feature), though I'm not sure which section would be appropriate

L3MON4D3 commented on lua/luasnip/config.lua:
```diff
@@ -188,10 +188,16 @@ c = {
-					.. (session.config.enable_autosnippets and [[
+					.. (
+						session.config.enable_autosnippets
+							-- Luasnip_just_inserted makes sure we only trigger autosnippets after we inserted characters in insert mode.
+							-- vim.b.Luasnip_last_changedtick allows us to compute how many changes occurred in the file between two TextChanged* events. We use this information to distinguish typing characters by hand vs using <C-r> to paste multiple characters in insert mode. We want to allow expanding autosnippets in the first case and not in the second. After some investigation, this is the only way I found to di ...
```
I'm not 100% happy with this comment, I think it could be more descriptive, ie:
1. what is the issue
2. how can typing characters and pasting multiple characters be distinguished
3. the solution

Also, could you make the comment 80-100 characters wide? Normally stylua should take care of this, but here it doesn't :shrug:

L3MON4D3 replied:
(Actually, maybe stylua isn't supposed to..)

L3MON4D3 replied:
Oh, and in a similar vein, could you shorten the comments in the tests?

L3MON4D3 commented on lua/luasnip/config.lua:
```diff
@@ -188,10 +188,16 @@ c = {
 			autocmd InsertCharPre * lua Luasnip_just_inserted = true
-			autocmd TextChangedI,TextChangedP * lua if Luasnip_just_inserted then require("luasnip").expand_auto() Luasnip_just_inserted=nil end
-		]] or "")
+			autocmd TextChangedI,TextChangedP * lua if Luasnip_just_inserted then Luasnip_just_inserted=nil if vim.b.Luasnip_last_changedtick == nil or vim.b.changedtick==vim.b.Luasnip_last_changedtick + 1 then require("luasnip").expand_auto() end vim.b.Luasnip_last_changedtick = vim.b.changedtick end
```
Similar here, could you spread this over multiple lines to increase readability?
Alternative would be toswitch to lua-autocommands, but I don't think this alone warrants it.

L3MON4D3 commented:
> Or put it into the doc (since it fixes an issue, and is not just a new feature), though I'm not sure which section would be appropriate

Actually, this behaviour might be unexpected, would you add a new section on it in `DOC.md`?

L3MON4D3 commented on tests/integration/add_snippets_spec.lua:
```diff
@@ -291,6 +303,22 @@ describe("add_snippets", function()
 
+		-- Test to make sure that autosnippets do not get triggered while pasting in insert mode.
+		feed("<ESC>dd") -- clear line
+		helpers.feed_command("set paste") -- disable autosnippets
```
Oh, why does `set paste` disable autosnippets? Had a cursory read of the help-section on it, but nothing in there caught my eye

L3MON4D3 commented:
Hey, are you still up for this? I could also take over, but I don't understand the `set paste`-part, I'd appreciate if you can explain that to me :sweat_smile:

leiserfg commented:
I don't get it either, also `paste` is deprecated it may stop working at any point.

L3MON4D3 commented:
Oh good to know, I guess we can ignore that then.
Still, important to know what's going on there, maybe it does not only happen when paste is set, but under other circumstances as well

carbon-steel commented:
Hi, sorry for dropping out from this PR. It slipped through the cracks when things in life started happening. I'll close this PR for now so as to not leave clutter behind and to signal that anyone should feel free to pick this work up for themselves.
